packages/graphql-types-generator/src/swift/appSyncCompatibilityTypes.ts

```diff
@@ -0,0 +1,419 @@
+// Blank lines at the beginning are intentional
+export const appSyncCompatibilityTypesCode = `
+import Foundation
+
+public protocol GraphQLInputValue {
+}
+
+public struct GraphQLVariable {
+  let name: String
+  
+  public init(_ name: String) {
+    self.name = name
+  }
+}
+
+extension GraphQLVariable: GraphQLInputValue {
+}
+
+extension JSONEncodable {
+  public func evaluate(with variables: [String: JSONEncodable]?) throws -> Any {
+    return jsonValue
+  }
+}
+
+public typealias GraphQLMap = [String: JSONEncodable?]
+
+extension Dictionary where Key == String, Value == JSONEncodable? {
+  public var withNilValuesRemoved: Dictionary<String, JSONEncodable> {
+    var filtered = Dictionary<String, JSONEncodable>(minimumCapacity: count)
+    for (key, value) in self {
+      if value != nil {
+        filtered[key] = value
+      }
+    }
+    return filtered
+  }
+}
+
+public protocol GraphQLMapConvertible: JSONEncodable {
+  var graphQLMap: GraphQLMap { get }
+}
+
+public extension GraphQLMapConvertible {
+  var jsonValue: Any {
+    return graphQLMap.withNilValuesRemoved.jsonValue
+  }
+}
+
+public typealias GraphQLID = String
+
+public protocol GraphQLOperation: AnyObject {
+  
+  static var operationString: String { get }
+  static var requestString: String { get }
+  static var operationIdentifier: String? { get }
+  
+  var variables: GraphQLMap? { get }
+  
+  associatedtype Data: GraphQLSelectionSet
+}
+
+public extension GraphQLOperation {
+  static var requestString: String {
+    return operationString
+  }
+
+  static var operationIdentifier: String? {
+    return nil
+  }
+
+  var variables: GraphQLMap? {
+    return nil
+  }
+}
+
+public protocol GraphQLQuery: GraphQLOperation {}
+
+public protocol GraphQLMutation: GraphQLOperation {}
+
+public protocol GraphQLSubscription: GraphQLOperation {}
+
+public protocol GraphQLFragment: GraphQLSelectionSet {
+  static var possibleTypes: [String] { get }
+}
+
+public typealias Snapshot = [String: Any?]
+
+public protocol GraphQLSelectionSet: Decodable {
+  static var selections: [GraphQLSelection] { get }
+  
+  var snapshot: Snapshot { get }
+  init(snapshot: Snapshot)
+}
+
+extension GraphQLSelectionSet {
+    public init(from decoder: Decoder) throws {
+        if let jsonObject = try? APISwiftJSONValue(from: decoder) {
+            let encoder = JSONEncoder()
+            let jsonData = try encoder.encode(jsonObject)
+            let decodedDictionary = try JSONSerialization.jsonObject(with: jsonData, options: []) as! [String: Any]
+            let optionalDictionary = decodedDictionary.mapValues { $0 as Any? }
+
+            self.init(snapshot: optionalDictionary)
+        } else {
+            self.init(snapshot: [:])
+        }
+    }
+}
```

Comment:
This is where the custom decoder was added. It uses an internally defined JSON value type called `APISwiftJSONValue` (to avoid developer defined "JSONValue" type) to get the data into an JSON object before serializing and deserializing back to the expected output a Snapshot (typealias for [String:Any?]))